### PR TITLE
BZX StakingAdapter

### PR DESCRIPTION
**Pull request by RomanHiden**

### CODE REVIEW

**Comment by sobolev-igor** [comment]
Hi @RomanHiden! Thanks for the PR! Do you have any example address for staking adapter and expected returned values, so we could validate everything is fine?

**Comment by sobolev-igor** [comment]
Also, if there are any planned updates for the staking pools, may be it's better to implement upgradable version of adapter?

**Comment by RomanHiden** [comment]
Thanks, I am not ale to reach you on discord for more productive discussion, i think due to privacy settings @sobolev-igor

File: contracts/adapters/bzx/BzxStakingAdapter.sol [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                StakingRewards(stakingContract).balanceOfByAsset(
+                    token,
+                    account
+                );
+        } else if (token == BPT) {
+            return
+                StakingRewards(stakingContract).balanceOfByAsset(
+                    token,
+                    account
+                );
```
Comment (sobolevigor986) [inline contracts/adapters/bzx/BzxStakingAdapter.sol:99]: all 3 cases can be combined into one if-case (`if (token == IBZRX || token == VBZRX || ... )`)

File: contracts/adapters/bzx/BzxStakingAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    string public constant override tokenType = "ERC20";
+
+    address internal constant IBZRX =
+        0x18240BD9C07fA6156Ce3F3f61921cC82b2619157;
+    address internal constant VBZRX =
+        0xB72B31907C1C95F3650b64b2469e08EdACeE5e8F;
+    address internal constant BZRX = 0x56d811088235F11C8920698a204A5010a788f4b3;
+    address internal constant BPT = 0xe26A220a341EAca116bDa64cF9D5638A935ae629;
+
+    address internal constant stakingContract =
```
Comment (sobolevigor986) [inline contracts/adapters/bzx/BzxStakingAdapter.sol:63]:
```suggestion
    address internal constant STAKING_CONTRACT =
```
as it's a constant

File: contracts/adapters/bzx/BzxVestingStakingAdapter.sol [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+ */
+contract BzxStakingAdapter is ProtocolAdapter {
+    string public constant override adapterType = "Asset";
+
+    string public constant override tokenType = "ERC20";
+
+    address internal constant IBZRX =
+        0x18240BD9C07fA6156Ce3F3f61921cC82b2619157;
+    address internal constant VBZRX =
+        0xB72B31907C1C95F3650b64b2469e08EdACeE5e8F;
```
Comment (sobolevigor986) [inline contracts/adapters/bzx/BzxVestingStakingAdapter.sol:53]: not used in this contract

File: contracts/adapters/bzx/BzxVestingStakingAdapter.sol [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    address internal constant BZRX = 0x56d811088235F11C8920698a204A5010a788f4b3;
+    address internal constant BPT = 0xe26A220a341EAca116bDa64cF9D5638A935ae629;
```
Comment (sobolevigor986) [inline contracts/adapters/bzx/BzxVestingStakingAdapter.sol:55]: not used in this contract

File: contracts/adapters/bzx/BzxStakingAdapter.sol [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        view
+        returns (uint256 balance);
+}
+
+/**
+ * @title Adapter for BZX protocol (staking).
+ * @dev Implementation of ProtocolAdapter interface.
+ * @author Roman Iftodi <romeo8881@gmail.com>
+ */
+contract BzxStakingAdapter is ProtocolAdapter {
```
Comment (sobolevigor986) [inline contracts/adapters/bzx/BzxStakingAdapter.sol:51]: Change name

File: contracts/adapters/bzx/BzxStakingAdapter.sol [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        );
+
+    function balanceOfByAsset(address token, address account)
+        external
+        view
+        returns (uint256 balance);
+}
+
+/**
+ * @title Adapter for BZX protocol (staking).
```
Comment (sobolevigor986) [inline contracts/adapters/bzx/BzxStakingAdapter.sol:47]: comment that this adapter used for vested rewards

**Review by sobolevigor986 — CHANGES_REQUESTED** [review]

File: contracts/adapters/bzx/BzxVestingStakingAdapter.sol [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    address internal constant CURVE3CRV =
+        0x6c3F90f043a72FA612cbac8115EE7e52BDe6E490;
+
+    address internal constant stakingContract =
```
Comment (sobolevigor986) [inline contracts/adapters/bzx/BzxVestingStakingAdapter.sol:59]:
```suggestion
    address internal constant STAKING_CONTRACT =
```
as it's a constant

File: contracts/adapters/bzx/BzxStakingAdapter.sol [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+            (, uint256 curve3crv, , ) =
+                StakingRewards(stakingContract).earned(account);
+            return
+                StakingRewards(stakingContract).balanceOfByAsset(
+                    token,
+                    account
+                ) + curve3crv;
+        } else {
+            return 0;
+        }
```
Comment (sobolevigor986) [inline contracts/adapters/bzx/BzxStakingAdapter.sol]:
```suggestion
        uint256 totalAmount = 0;
        if (token == IBZRX || token == VBZRX || token == BZRX || token == BPT) {
            totalAmount += StakingRewards(STAKING_CONTRACT).balanceOfByAsset(
                    token,
                    account
                );
        }
        if (token == BZRX) {
            (uint256 bzrxEarnings, , , ) =
                StakingRewards(STAKING_CONTRACT).earned(account);
            totalAmount += bzrxEarnings;
        }
        if (token == CURVE3CRV) {
            (, uint256 curve3crvEarnings, , ) =
                StakingRewards(STAKING_CONTRACT).earned(account);
            totalAmount +=  curve3crvEarnings;
        }
        return totalAmount;
```

**Review by sobolevigor986 — CHANGES_REQUESTED** [review]

File: contracts/adapters/bzx/BzxStakingAdapter.sol (hunk 5 above)
Comment (RomanHiden) [reply contracts/adapters/bzx/BzxStakingAdapter.sol:60]: I changed the other one to BzxVestingStakingRewards and this left unchanged

**Pull request merged**